File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    # 힙큐 = 최솟값부터
    # -642 -45 45 97 333 653
    # heapq[0]이 최솟값인건 신뢰할 수 있지만, heapq[-1]이 최댓값인건 신뢰할 수 없음
    answer = []
    min_heap = []
    max_heap = []
    visited = [False] * len(operations) # i번째 넣은 값이 유효한지
    
    for i, op in enumerate(operations):
        if op[0] == 'I':
            num = int(op[2:])
            heapq.heappush(min_heap, (num, i))
            heapq.heappush(max_heap, (-num, i))
            visited[i] = True
        elif op[0] == 'D':
            if op[2:] == '1':
                while max_heap and visited[max_heap[0][1]] == False:
                    heapq.heappop(max_heap)
                if max_heap:
                    visited[max_heap[0][1]] = False
                    heapq.heappop(max_heap)
            else:
                while min_heap and visited[min_heap[0][1]] == False:
                    heapq.heappop(min_heap)
                if min_heap:
                    visited[min_heap[0][1]] = False
                    heapq.heappop(min_heap)
                    
    while max_heap and visited[max_heap[0][1]] == False:
        heapq.heappop(max_heap)
    while min_heap and visited[min_heap[0][1]] == False:
        heapq.heappop(min_heap)
    
    if not max_heap:
        answer.append(0)
        answer.append(0)
    else:
        answer.append(-max_heap[0][0])
        answer.append(min_heap[0][0])
    
    return answer

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(vector<string> operations) {
    vector<int> answer(2,0);
    priority_queue<pair<int, int>> pq_max;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq_min;
    vector<bool> alive;
    int num_cnt = 0;
    int idx = 0;
    
    for (auto& s : operations){
        int num = stoi(s.substr(2));
        
        
        if (s[0] == 'I'){
            pq_max.push({num, idx});
            pq_min.push({num, idx});
            alive.push_back(true);
            idx++;
            num_cnt++;
        }
        else{
            if (num_cnt == 0){
                continue;
            }
            
            if (num == 1){
                while (!pq_max.empty() && !alive[pq_max.top().second]){
                    pq_max.pop();
                }
                if (!pq_max.empty()){
                    alive[pq_max.top().second] = false;
                    pq_max.pop();
                    num_cnt--;
                }
            }
            else{
                while (!pq_min.empty() && !alive[pq_min.top().second]){
                    pq_min.pop();
                }
                if (!pq_min.empty()){
                    alive[pq_min.top().second] = false;
                    pq_min.pop();
                    num_cnt--;
                }
            }
        }
    }
    
    if (num_cnt > 0){
        while (!pq_max.empty() && !alive[pq_max.top().second]) pq_max.pop();
        while (!pq_min.empty() && !alive[pq_min.top().second]) pq_min.pop();
        answer[0] = pq_max.top().first;
        answer[1] = pq_min.top().first;
    }
    
    return answer;
}